openmls/src/group/mls_group/commit_builder.rs

```diff
@@ -0,0 +1,795 @@
+            .group_context()
+            .tls_serialize_detached()
+            .map_err(LibraryError::missing_bound_check)?;
+
+        let welcome_secret = key_schedule
+            .welcome(crypto, builder.group.ciphersuite())
+            .map_err(|_| LibraryError::custom("Using the key schedule in the wrong state"))?;
+        key_schedule
+            .add_context(crypto, &serialized_provisional_group_context)
+            .map_err(|_| LibraryError::custom("Using the key schedule in the wrong state"))?;
+        let provisional_epoch_secrets = key_schedule
+            .epoch_secrets(crypto, builder.group.ciphersuite())
+            .map_err(|_| LibraryError::custom("Using the key schedule in the wrong state"))?;
+
+        // Calculate the confirmation tag
+        let confirmation_tag = provisional_epoch_secrets
+            .confirmation_key()
+            .tag(
+                crypto,
+                builder.group.ciphersuite(),
+                diff.group_context().confirmed_transcript_hash(),
+            )
+            .map_err(LibraryError::unexpected_crypto_error)?;
+
+        // Set the confirmation tag
+        authenticated_content.set_confirmation_tag(confirmation_tag.clone());
+
+        diff.update_interim_transcript_hash(ciphersuite, crypto, confirmation_tag.clone())?;
+
+        // only computes the group info if necessary
+        let group_info = if !apply_proposals_values.invitation_list.is_empty()
+            || builder.group.configuration().use_ratchet_tree_extension
+        {
+            // Create the ratchet tree extension if necessary
+            let external_pub = provisional_epoch_secrets
+                .external_secret()
+                .derive_external_keypair(crypto, ciphersuite)
+                .map_err(LibraryError::unexpected_crypto_error)?
+                .public;
+            let external_pub_extension =
+                Extension::ExternalPub(ExternalPubExtension::new(external_pub.into()));
+            let other_extensions: Extensions =
+                if builder.group.configuration().use_ratchet_tree_extension {
+                    Extensions::from_vec(vec![
+                        Extension::RatchetTree(RatchetTreeExtension::new(
+                            diff.export_ratchet_tree(),
+                        )),
+                        external_pub_extension,
+                    ])?
+                } else {
+                    Extensions::single(external_pub_extension)
+                };
+
+            // Create to-be-signed group info.
+            let group_info_tbs = {
+                GroupInfoTBS::new(
+                    diff.group_context().clone(),
+                    other_extensions,
+                    confirmation_tag,
+                    builder.group.own_leaf_index(),
+                )
+            };
+            // Sign to-be-signed group info.
+            Some(group_info_tbs.sign(signer)?)
+        } else {
+            None
+        };
```

Comment:
Just a nit (and I'm not sure if it was like this before), but I'd slightly prefer to invert the check and return `None` in the `if` and do the elaborate computation in the `else`. That way when reading the code I immediately see the obvious `None` return and don't have to parse the `else` without seeing the original condition. The same goes for the `welcome_option` below.

Reply:
It was like this before, but I think I agree with the general criticism. What I did now is that I assigned the if-conditions to variables and then do 

```rust
let group_info = if !needs_group_info {
  None
} else {
   ... ;
  Some(...)
};`